core/src/avm2/globals/array.rs

```diff
@@ -0,0 +1,1353 @@
+        QName::new(Namespace::public_namespace(), "pop"),
+        Method::from_builtin(pop),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "push"),
+        Method::from_builtin(push),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "reverse"),
+        Method::from_builtin(reverse),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "shift"),
+        Method::from_builtin(shift),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "unshift"),
+        Method::from_builtin(unshift),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "slice"),
+        Method::from_builtin(slice),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "splice"),
+        Method::from_builtin(splice),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "sort"),
+        Method::from_builtin(sort),
+    ));
+
+    class.write(mc).define_instance_trait(Trait::from_method(
+        QName::new(Namespace::public_namespace(), "sortOn"),
+        Method::from_builtin(sort_on),
+    ));
+
+    class.write(mc).define_class_trait(Trait::from_const(
+        QName::new(Namespace::public_namespace(), "CASEINSENSITIVE"),
+        Multiname::from(QName::new(Namespace::public_namespace(), "uint")),
+        Some(EnumSet::from(SortOptions::CaseInsensitive).as_u32().into()),
+    ));
+
+    class.write(mc).define_class_trait(Trait::from_const(
+        QName::new(Namespace::public_namespace(), "DESCENDING"),
+        Multiname::from(QName::new(Namespace::public_namespace(), "uint")),
+        Some(EnumSet::from(SortOptions::Descending).as_u32().into()),
+    ));
+
+    class.write(mc).define_class_trait(Trait::from_const(
+        QName::new(Namespace::public_namespace(), "NUMERIC"),
+        Multiname::from(QName::new(Namespace::public_namespace(), "uint")),
+        Some(EnumSet::from(SortOptions::Numeric).as_u32().into()),
+    ));
+
+    class.write(mc).define_class_trait(Trait::from_const(
+        QName::new(Namespace::public_namespace(), "RETURNINDEXEDARRAY"),
+        Multiname::from(QName::new(Namespace::public_namespace(), "uint")),
+        Some(
+            EnumSet::from(SortOptions::ReturnIndexedArray)
+                .as_u32()
+                .into(),
+        ),
+    ));
+
+    class.write(mc).define_class_trait(Trait::from_const(
+        QName::new(Namespace::public_namespace(), "UNIQUESORT"),
+        Multiname::from(QName::new(Namespace::public_namespace(), "uint")),
+        Some(EnumSet::from(SortOptions::UniqueSort).as_u32().into()),
+    ));
```

Comment:
Probably for a future PR as this one is already quite big, but these traits should be in the AS3 namespace. The methods of the core classes in AS3 are defiend both as instance traits (with the AS3 namespace) and on the prototype, see:
https://help.adobe.com/en_US/as3/learn/WS5b3ccc516d4fbf351e63e3d118a9b90204-7f3f.html#WS5b3ccc516d4fbf351e63e3d118a9b90204-7f17

This is also necessary for array to work when the variable is given a type, e.g.:
```
var a: Array = [];
a.push();
```
because the compiler emits the `findpropstrict` op with the specific AS3 namespace.

The class initializer for core classes does something like this to wire up the prototype:
```
      {
         prototype.valueOf = function():String
         {
            if(this === prototype)
            {
               return "";
            }
            var n:Namespace = this;
            return n.uri;
         };
         prototype.toString = function():String
         {
            if(this === prototype)
            {
               return "";
            }
            var n:Namespace = this;
            return n.toString();
         };
         _dontEnumPrototype(prototype);
      }
```

Reply:
So, after re-reading the Array documentation, it claims all methods except `toString`, `toLocaleString` and `valueOf` are scoped to the AS3 namespace. However, there's nothing in there about which methods are and aren't prototype-scoped. From tests I did on Animate CC 2020, `toString`, `toLocaleString`, and `valueOf` are all prototype-inherited and do not exist as traits, everything else has both trait and prototype definitions.

I originally special-cased `Object`'s prototype-based inheritance by having it make it's own prototype, but given the fact that other classes also do this, I probably should move all that into class initializers. However, we don't run those yet, so I'm going to hold those off for a separate "class initializers PR" that gets those run at the correct time.